infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -0,0 +1,296 @@
+        } catch (InterruptedException ex) {
+            log.error("Log Replication interrupted Exception: ", ex);
+        }
+    }
+
+    /**
+     * Consumer of the eventQueue.
+     * <p>
+     * This method consumes the log replication events and does the state transition.
+     */
+    private void consume() {
+        try {
+            if (state.getType() == LogReplicationRuntimeStateType.STOPPING) {
+                log.info("Log Replication Communication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            //  Block until an event shows up in the queue.
+            LogReplicationRuntimeEvent event = eventQueue.take();
+
+            try {
+                LogReplicationRuntimeState newState = state.processEvent(event);
+                transition(state, newState);
+                state = newState;
+            } catch (IllegalTransitionException illegalState) {
+                log.error("Illegal log replication event {} when in state {}", event.getType(), state.getType());
+            }
+
+            communicationFSMConsumer.submit(this::consume);
+
+        } catch (Throwable t) {
+            log.error("Error on event consumer: ", t);
+        }
+    }
+
+    /**
+     * Perform transition between states.
+     *
+     * @param from initial state
+     * @param to   final state
+     */
+    private void transition(LogReplicationRuntimeState from, LogReplicationRuntimeState to) {
+        log.trace("Transition from {} to {}", from, to);
+        from.onExit(to);
+        to.clear();
+        to.onEntry(from);
+    }
+
+    public synchronized void updateConnectedEndpoints(String endpoint) {
+        connectedEndpoints.add(endpoint);
+    }
+
+    public synchronized void updateDisconnectedEndpoints(String endpoint) {
+        connectedEndpoints.remove(endpoint);
+    }
+
+    public synchronized void setLeaderEndpoint(String leader) {
+        leaderEndpoint = Optional.ofNullable(leader);
+    }
+
+    public synchronized Optional<String> getLeader() {
+        return leaderEndpoint;
+    }
+
+    public synchronized Set<String> getConnectedEndpoints() {
+        return connectedEndpoints;
+    }
+
+    /**
+     * Retrieve total number of entries to be sent based on a given timestamp.
+     *
+     * This is required for progress status reporting.
+     *
+     * @param ts base (reference) timestamp
+     *
+     * @return pending number of entries to send
+     */
+    public long getNumEntriesToSend(long ts) {
+        long ackTS = sourceManager.getLogReplicationFSM().getAckedTimestamp();
+        return ts - ackTS;
+    }
+
+    /**
+     * Stop Log Replication, regardless of current state.
+     */
+    public void stop() {
+        log.info("Local leadership lost. Log Replication will immediately stop.");
+        input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.LOCAL_LEADER_LOSS));
+    }
+}
```

Comment:
It seems like LOCAL_LEADER_LOSS event can not stop log replication, so discovery service is not able to be shut down.

Reply:
I am reviewing the state machine, let me take special look at this case! I'll let you know!